Test circulation API

leonardr opened the pull request:
This branch fixes an edge case in renewals. Overdrive (and possibly other distributors) prohibits renewing a loan if there are people waiting in line for the book. If you try, the `OverdriveAPI` raises `NoAvailableCopies`. In `CirculationAPI` weren't prepared for this--we only prepared for `CannotRenew`, which is raised if there _are_ available copies but your loan isn't near enough to its expiry date for renewal to be an option.

This real improvement in this branch is `test_circulationapi.py`, which contains the first tests of the troublesome `CirculationAPI` class. I had put this off because properly testing the code requires two levels of mock objects--a mock `CirculationAPI` that calls mock remote API objects--but this is so important to the circ manager and there are so many weird edge cases that I finally had to do it. I've added tests for all the `borrow()` cases that end with an active loan for the book. There are a lot more tests to be written and I'll get to them as I can.

aslagle commented on api/circulation.py:
```diff
@@ -200,7 +200,28 @@ def parse_fines(fines):
+        )
+        
+        loan_info = None
+        hold_info = None
```
these two lines are duplicated now

leonardr commented:
I found a WHOLE OTHER SET of mock objects for this code lying unused and (slightly) broken in `circulation.py` I moved it to the test class, got it to work again, and rewrote the tests to use it. I also added two more tests to show the difference in how `NoAvailableCopies` is handled when you already have a loan (it's a renewal error) versus when you have no loan (it gets turned into a hold).

leonardr commented:
Turns out those mock objects weren't unused; they're actually used in test_controller.py, which contains some tests that really ought to be in test_circulationapi.py. They're not really testing the controller, it's just that testing the controller was the only convenient way to test features of CirculationAPI.

aslagle commented on api/testing.py:
```diff
@@ -0,0 +1,144 @@
+        return self._db.query(Loan).filter(Loan.patron==patron)
+
+    def local_holds(self, patron):
+        return self._db.query(Hold).filter(Loan.patron==patron)
```
The filter should be Hold.patron

aslagle commented:
Looks good except for that one copy/paste error
